<a href="https://colab.research.google.com/github/sadidoll/Artifical-Interlligence/blob/master/Homework_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import random as rnd
# visualization
import seaborn as sns

%matplotlib inline

Downloading Titanic Dataset

In [22]:
train_download = "https://docs.google.com/uc?export=download&id=1SxOK4zcBPL2C0ArlxFyLavnvEjgzjKQ_"
test_download = "https://docs.google.com/uc?export=download&id=1W6Lth3OEjzqyLg5O8U_TCT1oayebB2dE"

!wget -O train.csv --no-check-certificate "$train_download"
!wget -O test.csv --no-check-certificate "$test_download"

train_df = pd.read_csv('train.csv')
test_df = pd.read_csv('test.csv')

--2021-02-15 00:46:45--  https://docs.google.com/uc?export=download&id=1SxOK4zcBPL2C0ArlxFyLavnvEjgzjKQ_
Resolving docs.google.com (docs.google.com)... 172.217.3.14, 2607:f8b0:4026:801::200e
Connecting to docs.google.com (docs.google.com)|172.217.3.14|:443... connected.
HTTP request sent, awaiting response... 302 Moved Temporarily
Location: https://doc-00-68-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/g5ftr7ipumh15s5gecmqmf1hef30qvrt/1613349975000/16552832296577571015/*/1SxOK4zcBPL2C0ArlxFyLavnvEjgzjKQ_?e=download [following]
--2021-02-15 00:46:46--  https://doc-00-68-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/g5ftr7ipumh15s5gecmqmf1hef30qvrt/1613349975000/16552832296577571015/*/1SxOK4zcBPL2C0ArlxFyLavnvEjgzjKQ_?e=download
Resolving doc-00-68-docs.googleusercontent.com (doc-00-68-docs.googleusercontent.com)... 74.125.137.132, 2607:f8b0:4023:c03::84
Connecting to doc-00-68-docs.googleusercontent.com (doc-00-68-docs.googleuserc

# Can delete the info for null values at the end

In [4]:
# Displays the total null values for the training and test sets
print('Training set null values')
print(train_df.isnull().sum())

print('\nTest set null values')
print(test_df.isnull().sum())

Training set null values
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Test set null values
PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64


In [23]:
train_df.describe(include = ['O'])

Name   Sex  ... Cabin Embarked
count                                                 891   891  ...   204      889
unique                                                891     2  ...   147        3
top     Thorneycroft, Mrs. Percival (Florence Kate White)  male  ...    G6        S
freq                                                    1   577  ...     4      644

[4 rows x 5 columns]

###The rate of duplicates for the Ticket feature is:𝑁𝑢𝑚𝑏𝑒𝑟 𝑜𝑓 𝑑𝑢𝑝𝑙𝑖𝑐𝑎𝑡𝑒 𝑣𝑎𝑙𝑢𝑒𝑠 𝑖𝑛 𝑡𝑖𝑐𝑘𝑒𝑡 / 𝑇𝑜𝑡𝑎𝑙 𝑛𝑢𝑚𝑏𝑒𝑟 𝑜𝑓 𝑡𝑖𝑐𝑘𝑒𝑡 𝑣𝑎𝑙𝑢𝑒𝑠 = 210891 = 23%
By looking at the Ticket feature we can’t get a lot of information for our model. Also, Ticket feature has 681 unique values, and it will be difficult to convert the values into useful categories.The 'Ticket' feature does not correlate with the 'Survival' feature, so the 'Ticket' feature should be dropped. Aling with the Ticket feature PassengerID and Name features are also being dropped as they do not help in determining the survival

In [24]:
# Drops the 'Ticket' feature for the training and test sets, as well as the
# 'PassengerId' and 'Name' features, which doesn't help with determining
# survival
train_df_new = train_df.drop(['PassengerId', 'Name', 'Ticket'], axis=1)
test_df_new = test_df.drop(['PassengerId', 'Name', 'Ticket'], axis=1)


In [25]:
train_df_new

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Cabin Embarked
0           0       3    male  22.0      1      0   7.2500   NaN        S
1           1       1  female  38.0      1      0  71.2833   C85        C
2           1       3  female  26.0      0      0   7.9250   NaN        S
3           1       1  female  35.0      1      0  53.1000  C123        S
4           0       3    male  35.0      0      0   8.0500   NaN        S
..        ...     ...     ...   ...    ...    ...      ...   ...      ...
886         0       2    male  27.0      0      0  13.0000   NaN        S
887         1       1  female  19.0      0      0  30.0000   B42        S
888         0       3  female   NaN      1      2  23.4500   NaN        S
889         1       1    male  26.0      0      0  30.0000  C148        C
890         0       3    male  32.0      0      0   7.7500   NaN        Q

[891 rows x 9 columns]

In the training set the Cabin feature is not complete and have 687 null values. There are total of 1014 null values in the Cabin features of the combined dataset of training and test dataset. It would be better to drop the Cabin feature since 77% of the values of the Cabin feature are missing in the training dataset

In [9]:
test_df.describe(include = ['O'])

Name   Sex    Ticket            Cabin Embarked
count                             418   418       418               91      418
unique                            418     2       363               76        3
top     Peruschitz, Rev. Joseph Maria  male  PC 17608  B57 B59 B63 B66        S
freq                                1   266         5                3      270

In [10]:
train_df.describe(include = ['O'])

Name   Sex  ... Cabin Embarked
count                                                 891   891  ...   204      889
unique                                                891     2  ...   147        3
top     Thorneycroft, Mrs. Percival (Florence Kate White)  male  ...    G6        S
freq                                                    1   577  ...     4      644

[4 rows x 5 columns]

In [28]:
# Drops the 'Cabin' feature for the training and test sets
train_df_new.drop(['Cabin'], axis=1, inplace=True)
test_df_new.drop(['Cabin'], axis=1, inplace=True)

train_df_new

KeyError: ignored

Coverting Sex feature to a new feature called Gender where female=1 and male=0 (From previous homework).

In [29]:
# Creates a new dataframe with the feature of 'Sex' replaced with the feature of
# 'Gender', as well as 'male' replaced with '0' and 'female' replaced with '1'
train_df_new.rename({'Sex':'Gender'}, axis=1, inplace=True)
train_df_new['Gender'] = train_df_new['Gender'].replace(['male'],'0')
train_df_new['Gender'] = train_df_new['Gender'].replace(['female'],'1')

test_df_new.rename({'Sex':'Gender'}, axis=1, inplace=True)
test_df_new['Gender'] = test_df_new['Gender'].replace(['male'],'0')
test_df_new['Gender'] = test_df_new['Gender'].replace(['female'],'1')
train_df_new

Survived  Pclass Gender   Age  SibSp  Parch     Fare Embarked
0           0       3      0  22.0      1      0   7.2500        S
1           1       1      1  38.0      1      0  71.2833        C
2           1       3      1  26.0      0      0   7.9250        S
3           1       1      1  35.0      1      0  53.1000        S
4           0       3      0  35.0      0      0   8.0500        S
..        ...     ...    ...   ...    ...    ...      ...      ...
886         0       2      0  27.0      0      0  13.0000        S
887         1       1      1  19.0      0      0  30.0000        S
888         0       3      1   NaN      1      2  23.4500        S
889         1       1      0  26.0      0      0  30.0000        C
890         0       3      0  32.0      0      0   7.7500        Q

[891 rows x 8 columns]

A simple way is to generate random numbers between mean and standard deviation for estimating and completing features with missing or null values. We will first do this for the Age feature(From Homework1):

In [30]:
##data_age = [train_df]
# Combines the 'Age' feature of the training and test sets to get the mean and 
# median
age_df = train_df_new['Age'].append(test_df_new['Age'])
# Gets the mean and median of the 'Age' feature of the combined data sets
mean = age_df.mean()
median = age_df.median()

print('Age Mean: ' + str(mean))
print('Age Median: ' + str(median))

# Sets the min and max range of the random number generator
if mean > median:
  min = median
  max = mean
else:
  min = mean
  max = median

# Fills in the null values of the 'Age' feature in the training and test sets
# with random values between the mean and median
for i in range(train_df_new.index.stop):
  if np.isnan(train_df_new.loc[i].at['Age']):
    train_df_new.at[i,'Age'] = np.random.uniform(min, max)

for i in range(test_df_new.index.stop):
  if np.isnan(test_df_new.loc[i].at['Age']):
    test_df_new.at[i,'Age'] = np.random.uniform(min, max)

train_df_new

Age Mean: 29.881137667304014
Age Median: 28.0


Survived  Pclass Gender        Age  SibSp  Parch     Fare Embarked
0           0       3      0  22.000000      1      0   7.2500        S
1           1       1      1  38.000000      1      0  71.2833        C
2           1       3      1  26.000000      0      0   7.9250        S
3           1       1      1  35.000000      1      0  53.1000        S
4           0       3      0  35.000000      0      0   8.0500        S
..        ...     ...    ...        ...    ...    ...      ...      ...
886         0       2      0  27.000000      0      0  13.0000        S
887         1       1      1  19.000000      0      0  30.0000        S
888         0       3      1  29.219737      1      2  23.4500        S
889         1       1      0  26.000000      0      0  30.0000        C
890         0       3      0  32.000000      0      0   7.7500        Q

[891 rows x 8 columns]

In [31]:
# Combines the 'Embarked' feature of training and test sets to get the mode
embarked_df = train_df_new['Embarked'].append(test_df_new['Embarked'])

# Gets the mode of the 'Embarked' feature of the combined data sets
embarked_mode = embarked_df.mode()[0]
print('Embarked Mode: ' + str(embarked_mode))

# Replaces the null values with the mode of the 'Embarked' feature
train_df_new['Embarked'].fillna(embarked_mode, inplace=True)

train_df_new.head(63)

Embarked Mode: S


Survived  Pclass Gender   Age  SibSp  Parch     Fare Embarked
0          0       3      0  22.0      1      0   7.2500        S
1          1       1      1  38.0      1      0  71.2833        C
2          1       3      1  26.0      0      0   7.9250        S
3          1       1      1  35.0      1      0  53.1000        S
4          0       3      0  35.0      0      0   8.0500        S
..       ...     ...    ...   ...    ...    ...      ...      ...
58         1       2      1   5.0      1      2  27.7500        S
59         0       3      0  11.0      5      2  46.9000        S
60         0       3      0  22.0      0      0   7.2292        C
61         1       1      1  38.0      0      0  80.0000        S
62         0       1      0  45.0      1      0  83.4750        S

[63 rows x 8 columns]

In [32]:
# Combines the 'Fare' feature of training and test sets to get the mode
fare_df = train_df_new['Fare'].append(test_df_new['Fare'])

# Gets the mode of the 'Fare' feature of the combined data sets
fare_mode = fare_df.mode()[0]
print('Fare Mode: ' + str(fare_mode))

# Replaces the null values with the mode of the 'Fare' feature
test_df_new['Fare'].fillna(fare_mode, inplace=True)

test_df_new

Fare Mode: 8.05


Pclass Gender        Age  SibSp  Parch      Fare Embarked
0         3      0  34.500000      0      0    7.8292        Q
1         3      1  47.000000      1      0    7.0000        S
2         2      0  62.000000      0      0    9.6875        Q
3         3      0  27.000000      0      0    8.6625        S
4         3      1  22.000000      1      1   12.2875        S
..      ...    ...        ...    ...    ...       ...      ...
413       3      0  28.762625      0      0    8.0500        S
414       1      1  39.000000      0      0  108.9000        C
415       3      0  38.500000      0      0    7.2500        S
416       3      0  28.210168      0      0    8.0500        S
417       3      0  29.735518      1      1   22.3583        C

[418 rows x 7 columns]

In [33]:
# Replaces the value of the 'Fare' feature with oridinal bands for the training
# and test sets
for i in range(train_df_new.index.stop):
  fare = train_df_new.loc[i].at['Fare']
  if fare > -0.001 and fare <= 7.91:
    train_df_new.at[i,'Fare'] = 0
  elif fare > 7.91 and fare <= 14.454:
    train_df_new.at[i,'Fare'] = 1
  elif fare > 14.454 and fare <= 31.0:
    train_df_new.at[i,'Fare'] = 2
  elif fare > 31.0 and fare <= 512.329:
    train_df_new.at[i,'Fare'] = 3

for i in range(test_df_new.index.stop):
  fare = test_df_new.loc[i].at['Fare']
  if fare > -0.001 and fare <= 7.91:
    test_df_new.at[i,'Fare'] = 0
  elif fare > 7.91 and fare <= 14.454:
    test_df_new.at[i,'Fare'] = 1
  elif fare > 14.454 and fare <= 31.0:
    test_df_new.at[i,'Fare'] = 2
  elif fare > 31.0 and fare <= 512.329:
    test_df_new.at[i,'Fare'] = 3
  
train_df_new

Survived  Pclass Gender        Age  SibSp  Parch  Fare Embarked
0           0       3      0  22.000000      1      0   0.0        S
1           1       1      1  38.000000      1      0   3.0        C
2           1       3      1  26.000000      0      0   1.0        S
3           1       1      1  35.000000      1      0   3.0        S
4           0       3      0  35.000000      0      0   1.0        S
..        ...     ...    ...        ...    ...    ...   ...      ...
886         0       2      0  27.000000      0      0   1.0        S
887         1       1      1  19.000000      0      0   2.0        S
888         0       3      1  29.219737      1      2   2.0        S
889         1       1      0  26.000000      0      0   2.0        C
890         0       3      0  32.000000      0      0   0.0        Q

[891 rows x 8 columns]

In [37]:
print('\nTest set null values')
print(train_df_new.isnull().sum())


Test set null values
Survived    0
Pclass      0
Gender      0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64


In [34]:
train_df_new.corr()

Survived    Pclass       Age     SibSp     Parch      Fare
Survived  1.000000 -0.338481 -0.066743 -0.035322  0.081629  0.084540
Pclass   -0.338481  1.000000 -0.335685  0.083081  0.018443 -0.114414
Age      -0.066743 -0.335685  1.000000 -0.234112 -0.176338  0.028758
SibSp    -0.035322  0.083081 -0.234112  1.000000  0.414838 -0.012604
Parch     0.081629  0.018443 -0.176338  0.414838  1.000000  0.011364
Fare      0.084540 -0.114414  0.028758 -0.012604  0.011364  1.000000

Defining a function to calculate the Gini 

In [38]:
# Define function to calculate Gini Impurity
def get_gini_impurity(survived_count, total_count):
    survival_prob = survived_count/total_count
    not_survival_prob = (1 - survival_prob)
    random_observation_survived_prob = survival_prob
    random_observation_not_survived_prob = (1 - random_observation_survived_prob)
    mislabelling_survided_prob = not_survival_prob * random_observation_survived_prob
    mislabelling_not_survided_prob = survival_prob * random_observation_not_survived_prob
    gini_impurity = mislabelling_survided_prob + mislabelling_not_survided_prob
    return gini_impurity

In [41]:
test_df.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200

In [45]:
train_df_new[['Gender', 'Survived']].groupby(['Gender'], as_index=False).agg(['mean', 'count', 'sum'])

Survived           
            mean count  sum
Gender                     
0       0.188908   577  109
1       0.742038   314  233

In [47]:
gini_impurity = get_gini_impurity(342, 891)
gini_impurity

0.47301295786144265

In [49]:
# Gini for male
gini_impurity_male = get_gini_impurity(109, 577)
gini_impurity_male

0.3064437162277843

In [50]:
# Gini for Female
gini_impurity_male = get_gini_impurity(233, 314)
gini_impurity_male

0.3828350034484158